### PR TITLE
Convert text classification with switch transformer to Keras Core

### PR DESCRIPTION
**Pull request by freedomtan**
1. use keras_core to replace tf.keras, diff in 41bb533 
2. replace tf ops with `keras_core.ops` and implement a simple `unstack` because the is no `keras_core.ops.unstack`, diff in ee9b154

### CODE REVIEW

**Comment by freedomtan** [comment]
> Thanks for the PR. Did you run it with all backends? Do you have a Colab notebook?

It doesn't work for JAX and PyTorch. That's why I still put it in the `tensorflow` directory. I tried something like `KERAS_BACKEND=jax python text_classification_with_switch_transformer.py` and got
```
Using JAX backend.
25000 Training sequences
25000 Validation sequences
Number of tokens per batch: 10000
Traceback (most recent call last):
  File "/Users/freedom/work/keras-core/examples/keras_io/tensorflow/nlp/text_classification_with_switch_transformer.py", line 326, in <module>
    classifier = create_classifier()
  File "/Users/freedom/work/keras-core/examples/keras_io/tensorflow/nlp/text_classification_with_switch_transformer.py", line 294, in create_classifier
    x = transformer_block(x)
  File "/Users/freedom/work/keras-core/keras_core/utils/traceback_utils.py", line 123, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/Users/freedom/work/keras-core/keras_core/layers/layer.py", line 1097, in _maybe_build
    raise ValueError(
ValueError: Layer 'transformer_block' looks like it has unbuilt state, but Keras is not able to trace the layer `call()` in order to build it automatically. Possible causes:
1. The `call()` method of your layer may be crashing. Try to `__call__()` the layer eagerly on some test input first to see if it works. E.g. `x = np.random.random((3, 4)); y = layer(x)`
2. If the `call()` method is correct, then you may need to implement the `def build(self, input_shape)` method on your layer. It should create all variables used by the layer (e.g. by calling `layer.build()` on all its children layers).
```

[Colab notebook](https://colab.research.google.com/drive/1zoVFhheRYwdKlLTLWIGfOmRD0YJWK8zh?usp=sharing)

**Comment by fchollet** [comment]
Thanks! So since there's nothing in the code that's TF-specific, and it doesn't run in JAX or torch, I would classify that as a bug. The code looks like regular backend-agnostic code so I would expect to work with any backend. We should find out the root cause -- maybe we'll fix a bug in the framework that way. Try calling `keras.config.disable_traceback_filtering()` to get internal stack traces.

> ValueError: Layer 'transformer_block' looks like it has unbuilt state, but Keras is not able to trace the layer `call()` in order > to build it automatically. Possible causes:
> 1. The `call()` method of your layer may be crashing.

We should debug this eagerly (calling the layer on some concrete input and checking if that works)

**Comment by freedomtan** [comment]
I dug a bit into it. It seems the `batch size` for jax and torch is different.  I have a dummy example:
```python
import keras_core as keras
from keras_core import layers
from keras_core import ops


class DummyLayer(layers.Layer):
    def __init__(self):
        super().__init__()

    # Create the state of the layer (weights)
    def build(self, input_shape):
        print(input_shape)

    # Defines the computation
    def call(self, inputs):
        print(inputs, inputs.shape)
        return ops.convert_to_tensor(42, "int32")

# Instantiates the layer.
dummy_layer = DummyLayer()
inputs = layers.Input(shape=(200,200))
y = dummy_layer(inputs)
print(y)
``` 

Output
```sh
(tf-master) myway-m1:nlp freedom$ KERAS_BACKEND=tensorflow  python  /tmp/d.py
Using TensorFlow backend
(None, 200, 200)
Tensor("Placeholder:0", shape=(None, 200, 200), dtype=float32) (None, 200, 200)
<KerasTensor shape=(), dtype=int32, name=keras_tensor_1>
(tf-master) myway-m1:nlp freedom$ KERAS_BACKEND=jax  python  /tmp/d.py
Using JAX backend.
(None, 200, 200)
Traced<ShapedArray(float32[83,200,200])>with<DynamicJaxprTrace(level=1/0)> (83, 200, 200)
Traced<ShapedArray(float32[89,200,200])>with<DynamicJaxprTrace(level=1/0)> (89, 200, 200)
<KerasTensor shape=(), dtype=int32, name=keras_tensor_1>
(tf-master) myway-m1:nlp freedom$ KERAS_BACKEND=torch   python  /tmp/d.py
Using PyTorch backend.
(None, 200, 200)
tensor(..., device='meta', size=(83, 200, 200)) torch.Size([83, 200, 200])
tensor(..., device='meta', size=(89, 200, 200)) torch.Size([89, 200, 200])
<KerasTensor shape=(), dtype=int32, name=keras_tensor_1>

```

Is this expected behavior?

**Comment by fchollet** [comment]
> Is this expected behavior?

What you describe is in fact expected behavior. How does it relate to the issue with the example?

**Comment by freedomtan** [comment]
> > Is this expected behavior?
> 
> What you describe is in fact expected behavior. How does it relate to the issue with the example?

for this case, in [Switch.call()](https://github.com/freedomtan/keras-core/blob/convert_text_classification_with_switch_transformer/examples/keras_io/tensorflow/nlp/text_classification_with_switch_transformer.py#L216C13-L221)
```python
 def call(self, inputs):
        batch_size = ops.shape(inputs)[0]
        num_tokens_per_example = ops.shape(inputs)[1]

        # inputs shape: [num_tokens_per_batch, embed_dim]
        inputs = ops.reshape(inputs, [num_tokens_per_batch, self.embed_dim])
...
```
before running `inputs = ops.reshape(inputs, [num_tokens_per_batch, self.embed_dim])`

`num_tokens_per_batch = 10000`, `embed_dim = 32`
inputs.shape
- TF: (None, 200, 32)
- JAX: (83, 200, 32)

after running  `inputs = ops.reshape(inputs, [num_tokens_per_batch, self.embed_dim])`

TF: `inputs = ops.reshape(inputs, [num_tokens_per_batch, self.embed_dim])` reshaped input to `(10000, 32)` as expected
JAX: reshape failed

**Comment by fchollet** [comment]
I took a quick look, and it looks like there are multiple issues with the code.

- Make sure to call `super().__init__()` as the **first** statement in `__init__()` in custom layers. This needs to be fixed in 2 places.
- There are various unused examples / arguments in the example (e.g. `capacity_factor`)
- There is a shape mismatch in `Switch` at line `inputs = ops.reshape(inputs, [num_tokens_per_batch, self.embed_dim])`.

The last issue is because `num_tokens_per_batch` was never defined in the `call` method. Judging from the code, it needs to be defined as `batch_size * num_tokens_per_example`.

I think you need to rewrite the `Switch` class like:

```python
class Switch(layers.Layer):
    def __init__(self, num_experts, embed_dim):
        super().__init__()
        self.num_experts = num_experts
        self.embed_dim = embed_dim
        self.experts = [
            create_feedforward_network(embed_dim) for _ in range(num_experts)
        ]

    def build(self, input_shape):
        if None in input_shape:
            raise ValueError(
                "Expected a fully-defined input shape. "
                f"Received: input_shape={input_shape}"
            )
        batch_size = input_shape[0]
        num_tokens_per_example = input_shape[1]
        num_tokens_per_batch = batch_size * num_tokens_per_example
        expert_capacity = num_tokens_per_batch // self.num_experts
        self.router = Router(self.num_experts, expert_capacity)
        self.router.build(input_shape)

    def unstack(self, tensor, axis=0):
        shape = tensor.shape
        n = shape[axis]
        # output_shape = [shape[:axis]] + [n] + [shape[axis + 1:]]
        unpacked = ops.split(tensor, n, axis=axis)
        return unpacked

    def call(self, inputs):
        batch_size = ops.shape(inputs)[0]
        num_tokens_per_example = ops.shape(inputs)[1]
        num_tokens_per_batch = batch_size * num_tokens_per_example
        expert_capacity = num_tokens_per_batch // self.num_experts

        # inputs shape: [num_tokens_per_batch, embed_dim]
        inputs = ops.reshape(inputs, [num_tokens_per_batch, self.embed_dim])
        # dispatch_tensor shape: [expert_capacity, num_experts, tokens_per_batch]
        # combine_tensor shape: [tokens_per_batch, num_experts, expert_capacity]
        dispatch_tensor, combine_tensor = self.router(inputs)
        # expert_inputs shape: [num_experts, expert_capacity, embed_dim]
        expert_inputs = ops.einsum("ab,acd->cdb", inputs, dispatch_tensor)
        expert_inputs = ops.reshape(
            expert_inputs, [self.num_experts, expert_capacity, self.embed_dim]
        )
        # Dispatch to experts
        expert_input_list = self.unstack(expert_inputs, axis=0)
        expert_output_list = [
            self.experts[idx](expert_input)
            for idx, expert_input in enumerate(expert_input_list)
        ]
        # expert_outputs shape: [expert_capacity, num_experts, embed_dim]
        expert_outputs = ops.squeeze(ops.stack(expert_output_list, axis=2))
        # expert_outputs_combined shape: [tokens_per_batch, embed_dim]
        expert_outputs_combined = ops.einsum(
            "abc,xba->xc", expert_outputs, combine_tensor
        )
        # output shape: [batch_size, num_tokens_per_example, embed_dim]
        outputs = ops.reshape(
            expert_outputs_combined,
            [batch_size, num_tokens_per_example, self.embed_dim],
        )
        return outputs
```

Then you also need to add a `build()` method on the `Router` class. That's not all, but that's a start. The current code looks wrong in many places.

**Comment by freedomtan** [comment]
Thanks for taking time to check the code. It turned out as you pointed out the main issue for JAX backend is the batch size if not specified. With

```diff
diff --git a/examples/keras_io/tensorflow/nlp/text_classification_with_switch_transformer.py b/examples/keras_io/tensorflow/nlp/text_classification_with_switch_transformer.py
index 9a12cde..5061003 100644
--- a/examples/keras_io/tensorflow/nlp/text_classification_with_switch_transformer.py
+++ b/examples/keras_io/tensorflow/nlp/text_classification_with_switch_transformer.py
@@ -286,7 +286,7 @@ def create_classifier():
     switch = Switch(num_experts, embed_dim, num_tokens_per_batch)
     transformer_block = TransformerBlock(ff_dim, num_heads, switch)
 
-    inputs = layers.Input(shape=(num_tokens_per_example,))
+    inputs = layers.Input(shape=(num_tokens_per_example,), batch_size=batch_size)
     embedding_layer = TokenAndPositionEmbedding(
         num_tokens_per_example, vocab_size, embed_dim
     )

```

I can get expected training result with JAX backend.

**Comment by fchollet** [comment]
Nice -- does this fix work with torch as well?

**Comment by freedomtan** [comment]
> Nice -- does this fix work with torch as well?

unfortunately, nope, something wrong happened when using torch backend, e.g., in 'load_balanced_loss()', `ops.mean()` failed. I have to do  
```diff
diff --git a/examples/keras_io/tensorflow/nlp/text_classification_with_switch_transformer.py b/examples/keras_io/tensorflow/nlp/text_classification_with_switch_transformer.py
index c8cfafc..aa8403b 100644
--- a/examples/keras_io/tensorflow/nlp/text_classification_with_switch_transformer.py
+++ b/examples/keras_io/tensorflow/nlp/text_classification_with_switch_transformer.py
@@ -120,7 +120,7 @@ def load_balanced_loss(router_probs, expert_mask):
     # Want both vectors to have uniform allocation (1/num experts) across all
     # num_expert elements. The two vectors will be pushed towards uniform allocation
     # when the dot product is minimized.
-    loss = ops.mean(density_proxy * density) * ops.cast(
+    loss = ops.mean((density_proxy * density)[0], axis=0) * ops.cast(
         (num_experts**2), "float32"
     )
     return loss

```

**Comment by fchollet** [comment]
> ops.mean() failed

Why so? It is a bug with the torch backend? Could we fix it there?

**Comment by freedomtan** [comment]
> > ops.mean() failed
> 
> Why so? It is a bug with the torch backend? Could we fix it there?

Seems to be a bug of older PyTorch. I was using 1.12.0.
```python
import keras_core
from keras_core import ops


density_proxy = ops.convert_to_tensor([0.1557, 0.1633, 0.0530, 0.0914, 0.1284, 0.0655, 0.0174, 0.0645, 0.2147,
        0.0462])
density = ops.convert_to_tensor([[0., 0., 0., 0., 0., 0., 0., 0., 1., 0.]])
f = density_proxy * density

print(f"mean: {ops.mean(f[0], axis=0)}")
print(f"mean: {ops.mean(f)}")

```

The last line failed with Torch 1.12.0. After upgrade PyTorch to 2.0.1,  it works as expected.

**Comment by fchollet** [comment]
Got it. Normally we should require torch >= 2.0 so I think we're good. Is the PR ready to ready/merge?

**Comment by freedomtan** [comment]
> Got it. Normally we should require torch >= 2.0 so I think we're good. Is the PR ready to ready/merge?

I added one couple lines to resolve the last issue for Torch backend: its one-hot doesn't allow values to be larger than number of classes, 020129e

Thanks for your help.

File: examples/keras_io/nlp/text_classification_with_switch_transformer.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+        expert_mask *= ops.cast(
+            ops.less(
+                ops.cast(position_in_expert, "int32"), self.expert_capacity
+            ),
+            "float32",
+        )
+        expert_mask_flat = ops.sum(expert_mask, axis=-1)
+        # Mask out the experts that have overflowed the expert capacity.
+        expert_gate *= expert_mask_flat
+        # PyTorch one_hot doesn't allow values of position_in_expert to be
```
[the remark this answers is not in the document]
Comment (freedomtan) [reply examples/keras_io/nlp/text_classification_with_switch_transformer.py]: done in 9280ac3

File: keras_core/backend/torch/nn.py [hunk 2]
```diff
@@ -519,6 +519,9 @@ def one_hot(x, num_classes, axis=-1, dtype="float32"):
     # If axis is not last, change output to axis and shift remaining elements.
     x = convert_to_tensor(x, dtype=torch.long)
 
+    # Torch one_hot does not accept values >= num_classes
+    x = where(x < num_classes, x, 0)
```
Comment (fchollet) [inline keras_core/backend/torch/nn.py:523]: Please add a unit test to check this case (values >= num_classes)

Comment (freedomtan) [reply keras_core/backend/torch/nn.py:524]: done in b27adc3

**Comment by freedomtan** [comment]
huh, the np backend one_hot() caught the case I handled in torch backend. Fixed in 6cf8c45

**Comment by fchollet** [comment]
Keras Core is becoming Keras 3, and we're switching development to the main repository! If it is still relevant, please reopen this PR in the [keras-team/keras](https://github.com/keras-team/keras) repository. Unfortunately we aren't able to automatically transfer PRs (but we have transferred all issues).

**Pull request closed without merging**